_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,1079 @@
+   0x00000110 <+8>:	dsb	sy
+===> FAULT HERE
+   0x00000114 <+12>:	vmov	r0, s0
+   0x00000118 <+16>:	bx	lr
+```
+
+So we see there was a NOCP was due to issuing a FP instruction. The FPU is enabled using bits 20-23 of the [CPACR](https://interrupt.memfault.com/blog/cortex-m-rtos-context-switching#fpu-config-options) register located at `0xE000ED88`. A value of 0 indicates the extension is disabled. Let's check it:
+
+```
+(gdb) p/x (*(uint32_t*)0xE000ED88 >> 20) & 0xf
+$29 = 0x0
+```
+
+We can cleary see the FP Extension is disabled. We will have to enable the FPU to fix our bug.
+
+{: #imprecise-error-code-example}
+
+### Imprecise Fault
+
+#### Code
+
+```c
+void bad_addr_double_word_write(void) {
+  volatile uint64_t *buf = (volatile uint64_t *)0x30000000;
+  *buf = 0x1122334455667788;
+}
+```
+
+#### Analysis
+
+```
+(gdb) break main
+(gdb) continue
+Breakpoint 4, main () at ./cortex-m-fault-debug/main.c:182
+182	  xQueue = xQueueCreate(mainQUEUE_LENGTH, sizeof(unsigned long));
+(gdb) set g_crash_config=3
+(gdb) c
+Continuing.
+
+Program received signal SIGTRAP, Trace/breakpoint trap.
+0x0000021c in my_fault_handler_c (frame=0x200005e8 <ucHeap+1152>) at ./cortex-m-fault-debug/startup.c:91
+91	  HALT_IF_DEBUGGING();
+```
+
+Let's inspect `CFSR`:
+
+```
+(gdb) p/x *(uint32_t*)0xE000ED28
+$31 = 0x400
+(gdb) svd SCB CFSR_UFSR_BFSR_MMFSR
+Fields in SCB CFSR_UFSR_BFSR_MMFSR:
+[...]
+	IMPRECISERR:  1  Imprecise data bus error
+[...]
+```
+
+Yikes, the error is imprecise. This means the stack frame will point to the general area where the fault occurred but **not** the exact instruction!
+
+```
+(gdb) p/a *frame
+$32 = {
+  r0 = 0x55667788,
+  r1 = 0x11223344,
+  r2 = 0x10000000,
+  r3 = 0x30000000,
+  r12 = 0x200001b8 <ucHeap+80>,
+  lr = 0x199 <prvQueuePingTask+52>,
+  return_address = 0x198 <prvQueuePingTask+52>,
+  xpsr = 0x81000000
+}
+(gdb) list *0x198
+0x198 is in prvQueuePingTask (./cortex-m-fault-debug/main.c:162).
+157
+158	  while (1) {
+159	    vTaskDelayUntil(&xNextWakeTime, mainQUEUE_SEND_FREQUENCY_MS);
+160	    xQueueSend(xQueue, &ulValueToSend, 0U);
+161
+==> Crash somewhere around here
+162	    trigger_crash(g_crash_config);
+163	  }
+164	}
+165
+166	static void prvQueuePongTask(void *pvParameters) {
+```
+
+#### Analysis after making the Imprecise Error Precise
+
+If the crash was not readily reproducible we would have to inspect the code around this region and hypothesize what looks suspicious. However, [recall](#cortex-m3-m4-debug-trick) that there is a trick we can use for the Cortex-M4 to make all memory stores precise. Let's enable that and re-examine:
+
+```
+(gdb) mon reset
+Resetting target
+(gdb) c
+Continuing.
+
+Breakpoint 4, main () at ./cortex-m-fault-debug/main.c:182
+182	  xQueue = xQueueCreate(mainQUEUE_LENGTH, sizeof(unsigned long));
+(gdb) set g_crash_config=3
+
+==> Make all memory stores precise at the cost of performance
+(gdb) set *(uint32_t*)0xE000E008=(*(uint32_t*)0xE000E008 | 1<<1)
```

Comment:
Can you use SVD to set this instead of magic numbers?

Reply:
great idea, unfortunately the register isn't already in the svd files. I'm a bit sad arm doesn't maintain a set of these for core registers :/. I improved the comment instead!